File: MaWPictures/src/main/java/us/mikeandwan/photos/ui/BaseActivity.java
package us.mikeandwan.photos.ui;

import android.os.Bundle;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import us.mikeandwan.photos.MawApplication;
import us.mikeandwan.photos.di.ActivityModule;
import us.mikeandwan.photos.di.ApplicationComponent;


public class BaseActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        this.getApplicationComponent().inject(this);
    }


    protected ApplicationComponent getApplicationComponent() {
        return ((MawApplication)getApplication()).getApplicationComponent();
    }


    protected ActivityModule getActivityModule() {
        return new ActivityModule(this);
    }


    protected void updateToolbar(Toolbar toolbar, String title) {
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            ViewCompat.setElevation(toolbar, 4);

            if(title != null) {
                toolbar.setTitle(title);
            }
        }
    }
}

File: MaWPhotos/src/main/java/us/mikeandwan/photos/uiold/BaseActivity.kt
package us.mikeandwan.photos.uiold

import android.R.id.content
import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.ViewCompat
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.PublishSubject
import net.openid.appauth.AuthorizationException
import timber.log.Timber
import java.net.ConnectException
import java.util.concurrent.TimeUnit

@AndroidEntryPoint
@SuppressLint("Registered")
open class BaseActivity : AppCompatActivity() {
    private val _disposables = CompositeDisposable()
    private val _errorSubject = PublishSubject.create<String>()
    open fun onApiException(throwable: Throwable?) {
        handleApiException(throwable)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _disposables.add(_errorSubject
            .publish { publishedItems: Observable<String> ->
                publishedItems
                    .take(1)
                    .concatWith(
                        publishedItems
                            .skip(1)
                            .debounce(2, TimeUnit.SECONDS)
                    )
            }
            .subscribe { msg: String -> showError(msg) })
    }

    override fun onDestroy() {
        _disposables.clear()
        super.onDestroy()
    }

    protected fun updateToolbar(toolbar: Toolbar?, title: String?) {
        if (toolbar != null) {
            setSupportActionBar(toolbar)
            ViewCompat.setElevation(toolbar, 4f)
            if (title != null) {
                toolbar.title = title
            }
        }
    }

    protected fun handleApiException(throwable: Throwable?) {
        if (throwable == null) {
            return
        }
        Timber.e("Error accessing api: %s", throwable.message)
        if (throwable is ConnectException) {
            _errorSubject.onNext("Unable to connect to service at this time.")
        } else if (throwable is AuthorizationException) {
            _errorSubject.onNext("Authorization failed.")
        }
    }

    private fun showError(msg: String) {
        val view = findViewById<View>(content)
        Snackbar.make(view, msg, Snackbar.LENGTH_SHORT).show()
    }
}